// visua/dist/cssom/css-translate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dom_matrix_1 = require("./dom-matrix");
const css_1 = require("./css");
class CSSTranslate {
    constructor(x, y, z) {
        this.x = x;
        this.y = y;
        this.z = z;
        if (x.type.size !== 1 || !(x.type.has('length') || x.type.has('percent'))) {
            throw new TypeError(`Failed to construct CSSTranslate: ${x} is neither length nor percent`);
        }
        if (y.type.size !== 1 || !(y.type.has('length') || y.type.has('percent'))) {
            throw new TypeError(`Failed to construct CSSTranslate: ${y} is neither length nor percent`);
        }
        if (z && (z.type.size !== 1 || !z.type.has('length'))) {
            throw new TypeError(`Failed to construct CSSTranslate: ${z} is neither length nor percent`);
        }
        if (!z)
            this.z = css_1.CSS.px(0);
    }
    get is2D() {
        return this.z.value === 0;
    }
    toMatrix() {
        const x = this.x.to('px').value;
        const y = this.y.to('px').value;
        const z = this.z.to('px').value;
        return new dom_matrix_1.DOMMatrix().translateSelf(x, y, z);
    }
    toString() {
        let components = [this.x, this.y];
        if (!this.z.equals(css_1.CSS.px(0)))
            components.push(this.z);
        return `translate(${components.join(', ')})`;
    }
}
exports.CSSTranslate = CSSTranslate;
