// visua/dist/cssom/css-box-shadow-value.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const css_style_value_1 = require("./css-style-value");
const css_1 = require("./css");
class CSSShadow {
    constructor(components) {
        this.deafultLength = css_1.CSS.number(0);
        if (components.blurRadius != null && components.blurRadius.value < 0) {
            throw new TypeError('Failed to construct CSSShadow: Component blurRadius must be positive');
        }
        this.inset = components.inset || false;
        this.offsetX = components.offsetX || this.deafultLength;
        this.offsetY = components.offsetY || this.deafultLength;
        this.blurRadius = components.blurRadius || this.deafultLength;
        this.spreadDistance = components.spreadDistance || this.deafultLength;
        this.color = components.color;
    }
    toString() {
        let lengths = [this.offsetX, this.offsetY, this.blurRadius, this.spreadDistance]
            .filter(l => !l.equals(this.deafultLength));
        return `${this.inset ? 'inset ' : ''}${lengths.join(' ')} ${this.color}`;
    }
}
exports.CSSShadow = CSSShadow;
class CSSBoxShadowValue extends css_style_value_1.CSSStyleValue {
    constructor(layers) {
        super();
        this.layers = layers;
    }
    toString() {
        return this.layers.join(', ');
    }
}
exports.CSSBoxShadowValue = CSSBoxShadowValue;
